src/app/api/api.py

```diff
@@ -21,30 +26,87 @@
+    return request.app.state._db
+
+
 router = APIRouter()
 
+QUALITY_MATRIX_DESCRIPTION = """Calculation of the quality matrix.
+    For each replication source and each property, e.g., `cm:creator`, the quality matrix returns the ratio of
+    elements which miss this entry compared to the total number of entries.
+    A missing entry may be `cm:creator = null`.
+    Additional parameters:
+        store_to_db: Default False. Causes returned quality matrix to also be stored in the backend database."""
+TAG_STATISTICS = "Statistics"
+
 
 @router.get(
     "/quality_matrix",
     status_code=HTTP_200_OK,
     response_model=List[ColumnOutputModel],
     responses={HTTP_404_NOT_FOUND: {"description": "Quality matrix not determinable"}},
-    tags=["Statistics"],
-    description="""Calculation of the quality matrix.
-    For each replication source and each property, e.g., `cm:creator`, the quality matrix returns the ratio of
-    elements which miss this entry compared to the total number of entries.
-    A missing entry may be `cm:creator = null`.""",
+    tags=[TAG_STATISTICS],
+    description=QUALITY_MATRIX_DESCRIPTION,
+)
+async def get_quality_matrix(
+    database: Database = Depends(get_database), store_to_db=False
+):
+    _quality_matrix = await quality_matrix()
+    if store_to_db:
+        await stored_in_timeline(_quality_matrix, database)
+    return _quality_matrix
+
+
+@router.get(
+    "/quality_matrix/{timestamp}",
+    status_code=HTTP_200_OK,
+    response_model=List[ColumnOutputModel],
+    responses={HTTP_404_NOT_FOUND: {"description": "Quality matrix not determinable"}},
+    tags=[TAG_STATISTICS],
+    description=QUALITY_MATRIX_DESCRIPTION
+    + """A timestamp of the format XYZ yields the quality matrix at the respective date.""",
+)
+async def get_past_quality_matrix(
+    timestamp: Optional[int] = None, database: Database = Depends(get_database)
+):
+    if not timestamp:
+        return []
+
+    s = select([Timeline]).where(Timeline.timestamp == timestamp)
+    await database.connect()
+    result: list[Mapping[Timeline]] = await database.fetch_all(s)
+
+    if result is None:
+        raise HTTPException(status_code=404, detail="Item not found")
+    elif len(result) > 1:
+        raise HTTPException(status_code=500, detail="More than one item found")
+    return json.loads(result[0].quality_matrix)
+
+
+@router.get(
+    "/quality_matrix_timestamps",
+    status_code=HTTP_200_OK,
+    response_model=List[int],
+    responses={
+        HTTP_404_NOT_FOUND: {
+            "description": "Timestamps of old quality matrix results not determinable"
+        }
+    },
+    tags=[TAG_STATISTICS],
+    description="""Return timestamps of the format XYZ of past calculations of the quality matrix.""",
 )
-async def get_quality_matrix():
-    return await quality_matrix()
+async def get_timestamps(database: Database = Depends(get_database)):
+    return await timestamps(database)
```

Comment:
the other endpoints directly contain their implementation, but this one just redirects. Any particular reason for that? Otherwise, I would simply inline the timestamps function here to keep it consistent

Reply:
I thought about placing the implementation into separate functions, so you can test the function easily without calling the endpoint. So the question is, do we want easy tests (implementation in separate function) or code close to where it is called (implementation in endpoint, but difficult testing). I would prefer easy tests and therefore refactor the other endpoints to call specific functions. What do you think?